Docs for DockLayout (#160)
MicrosoftDocs/CommunityToolkit: docs/maui/layouts/DockLayout.md

---
title: DockLayout - .NET MAUI Community Toolkit
author: profix898
description: The DockLayout is a layout where views can be docked to the sides (top, left, right, bottom) of the layout container.
ms.date: 11/04/2022
---

# DockLayout

`DockLayout` is a layout where views can be docked to the sides of the layout container.

The image below shows how a `DockLayout` is conceptually structured. Child views are docked at one of 4 possible docking positions: *Top*, *Bottom*, *Left* or *Right* (equivalent to `DockPosition.Top`, `DockPosition.Bottom`, `DockPosition.Left`, and `DockPosition.Right`). Views that are not explicitly docked (or with `DockPosition.None`) are displayed at the center (or between *Top* / *Bottom* and *Left* / *Right* positions).

> ![DockLayout concept](/docs/images/layouts/DockLayout.svg)

Reviewer: Let's try a relative path

```suggestion
![DockLayout concept](../images/layouts/DockLayout.svg)
```

Reviewer: Ignore this I see you are already on the case

Author: The image was in a different 'images' folder. Moved it and updated to use relative path.

> ![DockLayout concept](./../../images/layouts/DockLayout.svg)

Author: Same thought here. Did that already ;)

             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
             x:Class="MyProject.MyContentPage">

    <mct:DockLayout>
        <Button mct:DockLayout.DockPosition="Top" Text="Top" HeightRequest="50" />
        <Button mct:DockLayout.DockPosition="Bottom" Text="Bottom" HeightRequest="70" />
        <Button mct:DockLayout.DockPosition="Left" Text="Left" WidthRequest="80" />
        <Button mct:DockLayout.DockPosition="Right" Text="Right" WidthRequest="90" />
        <Button Text="Center" />
    </mct:DockLayout>
    
</ContentPage>
```

For *Left* / *Right* docking, a `WidthRequest` should be specified. For *Top* / *Bottom* docking, a `HeightRequest` defines the size of the child view along the docking direction. The orthogonal directions are always calculated implicitly by the `DockLayout` manager.

### C#

To set the docking position from C#, use `DockLayout.SetDockPosition(IView, DockPosition)` to apply the attached `DockPosition` property.

```csharp
var button = new Button { Text = "Top", HeightRequest = 50 };
DockLayout.SetDockPosition(button, DockPosition.Top);
```

> Alternatively, `DockLayout` contains a fluent-style `Add(IView, DockPosition)` method that adds a child view to the layout container and sets it's docking position.

Reviewer: ```suggestion
Alternatively, `DockLayout` contains a fluent-style `Add(IView, DockPosition)` method that adds a child view to the layout container and sets its docking position.
```

```csharp
var dockLayout = new DockLayout();

> var button = dockLayout.Add(new Button { Text = "Top", HeightRequest = 50 }, DockPosition.Top);

Reviewer 2: ```suggestion
dockLayout.Add(new Button { Text = "Top", HeightRequest = 50 }, DockPosition.Top);
```

Reviewer: Is the `Add` method no longer fluent?

Reviewer 2: yes, it returns void. We can later change it if necessary, it won't be a breaking change

Author: Not sure I get why you made that change?

But at least then you should also change the text accordingly. It says `... contains a fluent-style ...`. Also, the C# code example is no longer useful, it should be changed to

```
var dockLayout = new DockLayout();
var button = new Button { Text = "Top", HeightRequest = 50 };
dockLayout.Add(button, DockPosition.Top);
```

Reviewer: I think I may have kickstarted the discussion that resulted in the fluent nature being removed. I got stuck in expecting that the Add would return the DockLayout rather than the newly created child so Adds could be chained together. I suspect we have taken the safest option in not deciding what it should return in the short term as introducing it isn't a breaking change but to remove or change it is.

I will make an effort to remove/update the section about the fluent nature.

Author: Thanks for the explanation. Actually my use case was a little different. Usually, after adding a view to the layout, I want to do something with it: e.g. set a binding to another view OR register an event (button pressed) OR the like. But I do see your use case as well and why the "ambiguity" might be an issue ...

Reviewer: You might want to checkout our Markup package and it's `Assign` method as it could cover that scenario: https://learn.microsoft.com/en-gb/dotnet/communitytoolkit/maui/markup/extensions/object-extensions#assign

## Customizing a DockLayout

A `DockLayout` container supports arbitrary `Padding` as well as several `DockLayout`-specific properties for customization. An example in XAML with all available options is given here:

### XAML

```xml
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
             x:Class="MyProject.MyContentPage">

    <mct:DockLayout HeightRequest="400"
                    WidthRequest="600"
                    Padding="10,20,30,40"
                    VerticalSpacing="10"
                    HorizontalSpacing="15"
                    ShouldExpandLastChild="False">
        ...
    </mct:DockLayout>
    
</ContentPage>
```

## Properties

|Property  |Type  |Description  |
|---------|---------|---------|
| `Padding` | `Thickness` | Gets or sets the padding around the layout container (inherited from `Layout`). |
| `HorizontalSpacing` | `double` | Gets or sets the **horizontal** spacing between docked views. |
| `VerticalSpacing` | `double` | Gets or sets the **vertical** spacing between docked views. |

> | <td colspan=2>`HorizontalSpacing` and `VerticalSpacing` is applied between neighboring views in the `DockLayout`. For example, `HorizontalSpacing` is added between *Left*, *None*, and *Right* views, but also between neighboring views in the same `DockPosition` such as multiple views docked to the *Left*. `VerticalSpacing` is rendered between vertically stacked views in *Top*, *None*, and *Bottom* positions.</td> |

Reviewer: I'll have to defer to @jfversluis here and let us know how this renders in the preview pages

Reviewer 3: Table looks good, the image in the top seems to be broken though: 
![image](https://user-images.githubusercontent.com/939291/200332531-40d53c5b-319c-46a2-8b55-398948ccf1f8.png)

Reviewer: Sorry can you check the preview again, I believe @profix898 has tried using a relative path to the image now

---
title: DockLayout - .NET MAUI Community Toolkit
author: profix898

> description: The `DockLayout` is a layout where views can be docked to the sides (top, left, right, bottom) of the layout container.

Reviewer: The metadata doesn't adhere to any markdown formatting

```suggestion
description: The DockLayout is a layout where views can be docked to the sides (top, left, right, bottom) of the layout container.
```

![DockLayout concept](/docs/images/layouts/DockLayout.svg)

## Building a DockLayout

A basic `DockLayout` can be created in XAML as shown here:

### XAML

```xml
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
             x:Class="MyProject.MyContentPage">

    <mct:DockLayout>
        <Button mct:DockLayout.DockPosition="Top" Text="Top" HeightRequest="50" />
        <Button mct:DockLayout.DockPosition="Bottom" Text="Bottom" HeightRequest="70" />
        <Button mct:DockLayout.DockPosition="Left" Text="Left" WidthRequest="80" />
        <Button mct:DockLayout.DockPosition="Right" Text="Right" WidthRequest="90" />
        <Button Text="Center" />
    </mct:DockLayout>
    
</ContentPage>
```

> For *Left* / *Right* docking, a `WidthRequest` should be specified. For *Top* / *Bottom* docking, a `HeightRequest` defines the size of the child view along the docking direction. The orthogonal directions are always calculated implictly by the `DockLayout` manager.

Reviewer: typo

```suggestion
For *Left* / *Right* docking, a `WidthRequest` should be specified. For *Top* / *Bottom* docking, a `HeightRequest` defines the size of the child view along the docking direction. The orthogonal directions are always calculated implicitly by the `DockLayout` manager.
```

> For *Left* / *Right* docking, a `WidthRequest` should be specified. For *Top* / *Bottom* docking, a `HeightRequest` defines the size of the child view along the docking direction. The orthogonal directions are always calculated implictly by the `DockLayout` manager.

Reviewer: Just to confirm is it a should or a must, what happens if no `WidthRequest` is supplied for `DockPosition.Left`?

Author: Should (not must): The `LayoutManager` uses the `DesiredSize` of each child view. In case no Height/WidthRequest is set, the view still returns a minimal size it wants to occupy.

### XAML

```xml
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
             x:Class="MyProject.MyContentPage">

    <mct:DockLayout HeightRequest="400"
                    WidthRequest="600"
                    Padding="10,20,30,40"
                    VerticalSpacing="10"
                    HorizontalSpacing="15"
                    ShouldExpandLastChild="False">
        ...
    </mct:DockLayout>
    
</ContentPage>
```

## Properties

|Property  |Type  |Description  |
|---------|---------|---------|
| `Padding` | `Thickness` | Gets or sets the padding around the layout container (inherited from `Layout`). |
| `VerticalSpacing` | `double` | Gets or sets the **vertical** spacing between docked views. |
| `HorizontalSpacing` | `double` | Gets or sets the **horizontal** spacing between docked views. |
| `ShouldExpandLastChild` | `bool` | If true, the last child is expanded to fill the remaining space (default: `true`). |

> `VerticalSpacing` and `HorizontalSpacing` is applied between neighboring views in the `DockLayout`. For example, `HorizontalSpacing` is added between *Left*, *None*, and *Right* views, but also between neighboring views in the same `DockPosition` such as multiple views docked to the *Left*. `VerticalSpacing` is rendered between vertically stacked views in *Top*, *None*, and *Bottom* positions.

Reviewer: I wonder if this information would be better included in the table above with the properties they directly relate to. What do you think?

Author: Feel free to relocate this section as you see fit. I like to keep (such) tables simple, so that you can get an overview of all properties at one glance. And better put detailed explanations below. But that's certainly a personal preference.

Author: Ok, I played a little with the table and tried to inline details on Horizontal/VerticalSpacing. Last (and separate) commit. Can be reverted easily, if you dont like it.
But as I said, feel free to change as you like best ...

Reviewer: Awesome thanks, as per my latest comment, let's see what Gerald notes on the preview of the actual docs page to make sure it handles this formatting.

## Additional Notes

> If `DockLayout` is used in a spatially constrainted place (especially with a size specified via `HeightRequest` or `WidthRequest` on the container), precedence is given by the order in which the child views are added to the `DockLayout` container. Consequently, whenever there is not enough space for all child views to be rendered, the least precedented ones (which were added last) will be removed upon layouting. For that reason, you should always check that the size of the container covers at least the minimum size of all its child views.

Reviewer: Some typos

```suggestion
If `DockLayout` is used in a spatially constrained place (especially with a size specified via `HeightRequest` or `WidthRequest` on the container), precedence is given by the order in which the child views are added to the `DockLayout` container. Consequently, whenever there is not enough space for all child views to be rendered, the least precedented ones (which were added last) will be removed upon rendering. For that reason, you should always check that the size of the container covers at least the minimum size of all its child views.
```

## Examples

> You can find an example of the `DockLayout` feature in action in the [.NET MAUI Community Toolkit Sample Application](https://github.com/CommunityToolkit/Maui/blob/main/samples/CommunityToolkit.Maui.Sample/Pages/Layouts/).

Reviewer: We should like direct to the DockLayoutPage here
```suggestion
You can find an example of the `DockLayout` feature in action in the [.NET MAUI Community Toolkit Sample Application](https://github.com/CommunityToolkit/Maui/blob/main/samples/CommunityToolkit.Maui.Sample/Pages/Layouts/DockLayoutPage.xaml).
```
